src/main/java/com/commercetools/sync/products/helpers/BatchProcessor.java

```diff
@@ -0,0 +1,273 @@
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import static com.commercetools.sync.products.helpers.VariantReferenceResolver.REFERENCE_ID_FIELD;
+import static com.commercetools.sync.products.helpers.VariantReferenceResolver.isProductReference;
+import static java.lang.String.format;
+import static java.util.Collections.emptySet;
+import static java.util.Objects.requireNonNull;
+import static java.util.Optional.empty;
+import static java.util.Optional.ofNullable;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+public class BatchProcessor {
+    static final String PRODUCT_DRAFT_KEY_NOT_SET = "ProductDraft with name: %s doesn't have a key. "
+        + "Please make sure all product drafts have keys.";
+    static final String PRODUCT_DRAFT_IS_NULL = "ProductDraft is null.";
+    static final String PRODUCT_VARIANT_DRAFT_IS_NULL = "ProductVariantDraft at position '%d' of ProductDraft "
+        + "with key '%s' is null.";
+    static final String PRODUCT_VARIANT_DRAFT_SKU_NOT_SET = "ProductVariantDraft at position '%d' of "
+        + "ProductDraft with key '%s' has no SKU set. Please make sure all variants have SKUs.";
+    static final String PRODUCT_VARIANT_DRAFT_KEY_NOT_SET = "ProductVariantDraft at position '%d' of "
+        + "ProductDraft with key '%s' has no key set. Please make sure all variants have keys.";
+
+    private final List<ProductDraft> productDrafts;
+    private final ProductSync productSync;
+    private final Set<ProductDraft> validDrafts = new HashSet<>();
+    private final Set<String> keysToCache = new HashSet<>();
+
+    public BatchProcessor(@Nonnull final List<ProductDraft> productDrafts, @Nonnull final ProductSync productSync) {
+        this.productDrafts = productDrafts;
+        this.productSync = productSync;
+    }
+
+    /**
+     * This method validates the batch of drafts, and only for valid drafts it adds the valid draft
+     * to {@code validDrafts} set, and adds the keys of all referenced products to
+     * {@code keysToCache}.
+     *
+     *
+     * <p>A valid product draft is one which satisfies the following conditions:
+     * <ol>
+     * <li>It has a key which is not blank (null/empty)</li>
+     * <li>It has all variants AND master variant valid</li>
+     * <li>A variant is valid if it satisfies the following conditions:
+     * <ol>
+     * <li>It has a key which is not blank (null/empty)</li>
+     * <li>It has a SKU which is not blank (null/empty)</li>
+     * </ol>
+     * </li>
+     * </ol>
+     */
+    public void validateBatch() {
```

Comment:
> For validation, I would expect it to inform me somehow about what is wrong by throwing error or return result.
but that's exactly what it does with `handleError` calls (if there are errors). It calls the sync process' error callback when there are errors. It additionally, collects keys and valid drafts into separate collections. But it's hard to put all this into one name. That's why I thought `validateBatch` and the JavaDoc would be enough to show what it does.  It could be `processBatch` maybe but I think it's too generic. What do you think? any suggestions?

Reply:
I think for splitting the name `partition` is good, although a bit unconventional. So `partitionBatchByValidity()`? 😄 
https://lodash.com/docs/4.17.4#partition